### DISCUSSION TITLE
Why powershell?

### DISCUSSION THREAD

**Discussion by donkus-schlonkus** [PowerShell/PowerShell#15023] [General]
I'm just kind of curious, when Linux has the terminal why want powershell? Isn't a ZSH or BASH terminal better?

**Comment by powercode** [comment]
There are pros and cons. If ZSH and BASH do all that you want, then keep using them. They are just tools, and hopefully not religions.

For example, the existing shells most likely have a better startup time than pwsh.
On the flip side, you can do things in PowerShell that the other shells can't, due to the different basic premise of using a pipeline of text instead of objects.

My experience is that I often spend time in the traditional shells expressing how to get something, while I in PowerShell just express what I want.

PowerShell is strong when it comes to creating structure, such as classes for output. It is also very easy to create functions with advanced parameter parsing and argument completion. For example

```powershell
function Get-LongRunningProcess {
[OutputType([system.diagnostics.Process])]
param(
  [Parameter(Mandatory)]
  [datetime] $OlderThan,
  [Parameter()]
  [ArgumentCompleter({param($c,$p,$w,$a,$f) get-process -Name $w* | Foreach-Object -MemberName Name})]
  $Include = '*'
)

  Get-Process -Name $Include | Where-Object StartTime -lt $OlderThan
}
```
When this function is defined, (it finds processes that have a StartTime older than a specific time), I can easily use it in another pipeline, with all structure maintained. It will complete parameter names, and process names for the `-Include` parameter. It will not allow the function to be called without a parameter for `$OlderThan`, and will ensure that it is a valid `datetime`.

```powershell
# Get processes that were not started today, where the first letter of the process name was in XYZ or with a WorkingSet greater than 100Mb
Get-LongRunningProcess ([datetime]::Now.Date) 
| Where { $_.Name[0] -in 'X', 'Y','Z' -or $_.WorkingSet -gt 100Mb }
```

This is possible due to the fact that objects, not strings, are flowing through the pipeline. This allows us to filter, sort, group, etc, by properties on the objects, or by mini-programs, ScriptBlocks as they are called in PowerShell. 
In combination, these features allow PowerShell to get closer to the unix ideal of small programs that does one thing, and does that well, and that can then be combined into solving bigger problems.

So, PowerShell has capabilities that I have not seen in any other shell. But it comes with a price and a part of that price is startup time.
I come from a bash background. My experience is that there is a bit of a learning curve with PowerShell, and it can be frustrating when you have preconceptions from other shells that make you think it works in a different way than it actually does. But once I got over the initial hurdles, I have never looked back. 
I find it to be, by far, the most powerful shell out there. 

Still, it's just a tool. Use whatever works for you.

**Comment by jahio** [comment]
I'm in the process of writing an article on PowerShell and I'll be offering an answer to this question in that article. I fully second what @powercode said above.

> Why PowerShell when we have `bash`, `zsh`, etc.?

The same reason that you have Coke, Sprite, and Dr. Pepper when you also have Pepsi: different people like different things. (But Coke is the best, just sayin' :trollface: )

Personally I found the distinction between _strings_ versus _objects_ by far the biggest reason to use PowerShell. I absolutely _hate_ the unmaintainable mess of regex filth that has to be created to extract data from STDOUT strings in typical shell land. If you've gotta pipe stuff through sed/awk/grep and manipulate text to get the information you want, that should be red flag number one that *something's wrong here*. Sadly, in a lot of cases there's just no alternative, no way to get _data_ totally independent of _formattng_ without doing all that crap.

So for me, that's where PowerShell really shines. It allows me to fetch raw data in some way, then pass just the data to other cmdlets, functions, or programs - including *NIX stuff like grep, awk, sed, and so on - to dramatically simplify my workflow.

Basically, the reason that output formatting for various *nix tools hasn't changed in decades - if ever - is that the shell scripts that glue the entire internet together would immediately crash overnight if output from something like, say, `ps` was substantially changed. Thousands of sysadmins would rage in unison as the world erupted into chaos and flames, because now they've got to go back and look at that shell script regex somebody else wrote 25 years ago that's been working ever since and fiddle with it to get the data out again so as not to break arguments to the next program on the other side of the shell redirection character.

Not with PowerShell. Formatting is totally separate from data, and you can even have full-fledged objects with properties and methods on top of that too. That - to me - is _totally_ worth the learning curve and work necessary to adapt my workflow to use `pwsh` day-to-day over `zsh` or `bash`. The fact that it can do lots of other cool stuff too is just icing on the cake.

> tl;dr

* ZSH/Bash: All output is a string and YOU have to extract the relevant parts with a shell script that is brittle because if the output changes upstream, it will break 99% of the time.
* PowerShell: Data is separate from output entirely; classes, objects, properties, methods, data types all supported natively out of the box. No need for brittle hacky shell crap, just send the _object_ down the line instead of the _string_. Way more stable, maintainable, far less prone to accidental/unexpected breakage.

**Comment by doctordns** [comment]
There are many more benefits of Powershell:

1. Great help. With man, you have to know what you are looking for. Plus, help is open source and is easy to improve.
2. Great command naming. To most,the (standard) verb and meaningful noun makes so much sense. 
3. Reflection. Not sure what is inside of objects returned by a cmdlet? Just pipe it to Get-member.
4. Object formatting by default. You can easily get output, and Powershell can format it easily.  Linux commands just give you the output the developer decides you should see as a string.
5. The sacred vow. Its just not available with Linux commands.

These are some of the benefits I rely on!

**Reply by jahio** [reply]
> The sacred vow. Its just not available with Linux commands.

Could you elaborate on this? I'm not sure what "the sacred vow" is here (probably because I'm pretty new to `pwsh`).

(And I agree entirely with the other points you mentioned!)
